### PR TITLE
mandelbrot: rayon and improvements

### PR DESCRIPTION
**Pull request by cristicbz**
This shaves off a consistent 7-10% on my machine, I'd be eager to see if you can reproduce these results. (though this wouldn't put us ahead of any other benchmarks, it will put us at under 2s on Isaac's machine).

### CODE REVIEW

**Comment by TeXitoi** [comment]
It doesn't compile on travis.

File: src/mandelbrot.rs [hunk 1]
```diff
@@ -50,28 +47,73 @@ impl Sub for Vecf64 {
...
-    let mut zi = ZEROS;
-    let mut tr = ZEROS;
-    let mut ti = ZEROS;
-    for _ in 0..MAX_ITER / 5 {
-        for _ in 0..5 {
-            zi = (zr + zr) * zi + ci;
-            zr = tr - ti + cr;
-            tr = zr * zr;
-            ti = zi * zi;
+#[derive(Default)]
```
Comment (TeXitoi) [inline src/mandelbrot.rs]: Removing this line fix compilation.

**Comment by TeXitoi** [comment]
| version | real | user |
|:--------:|:---:|:----:|
| before | 0m1.264s | 0m9.924s |
|  after  | 0m1.215s | 0m9.552s|
|  improvement  | 3.8% | 3.7% |

```
$ grep CPU /proc/cpuinfo | uniq -c
      8 model name	: Intel(R) Core(TM) i7-4702HQ CPU @ 2.20GHz
```

File: src/mandelbrot.rs [hunk 2]
```diff
@@ -50,28 +47,73 @@ impl Sub for Vecf64 {
...
+
+    pub fn run(&mut self, cr: Vecf64, cr2: Vecf64) -> u8 {
+        self.zr = cr;
+        self.zi = self.ci;
+        self.tr = cr2;
+        self.ti = self.ci2;
+        self.cr = cr;
+
+        self.advance(4);
+        for _ in 0..MAX_ITER / 5 - 1 {
```
Comment (TeXitoi) [inline src/mandelbrot.rs:84]: The fastest C version check every 7 iterations. Maybe that can improve the performances.

File: src/mandelbrot.rs (hunk 1 above)
Comment (cristicbz) [reply src/mandelbrot.rs]: Argh, sorry.

**Comment by cristicbz** [comment]
Right, I'm seeing a drop from 1.242s to 1.164s, which is actually only 6.7% not 7-10% like I said (it was late :P). I tried a whole bunch of options for the iteration counts with no luck. I guess this is one of the ones where without simd we won't do much better.

I liked the rayon version since it gets rid of the `collect` calls, and I packaged up the `Mandelbrot8` type to avoid repetition of the `advance` calls when optimizing the first run. Happy for you to close this if you don't like the changes subjectively, since the perf improvements aren't that significant.

**Comment by TeXitoi** [comment]
I think that rayon is more idiomatic, so I want to merge.

File: src/mandelbrot.rs [hunk 3]
```diff
@@ -3,23 +3,22 @@
...
 
 const MAX_ITER: usize = 50;
 const VLEN: usize = 8;
 const ZEROS: Vecf64 = Vecf64([0.; VLEN]);
 
 macro_rules! for_vec {
     ( in_each [ $( $val:tt ),* ] do $from:ident $op:tt $other:ident ) => {
-        $( $from.0[$val] $op $other.0[$val]; )*
+        let mut i = 0;
+        $( $from.0[i] $op $other.0[i]; i += 1; )*
```
Comment (TeXitoi) [inline src/mandelbrot.rs]: why do you change that?

File: src/mandelbrot.rs [hunk 4]
```diff
@@ -31,47 +30,95 @@ pub struct Vecf64([f64; VLEN]);
 impl Mul for Vecf64 {
     type Output = Vecf64;
     fn mul(mut self, other: Vecf64) -> Vecf64 {
-        for_vec!(self *= other);
+        for (s, o) in self.0.iter_mut().zip(&other.0) {
+            *s *= *o;
+        }
```
Comment (TeXitoi) [inline src/mandelbrot.rs]: why?

File: src/mandelbrot.rs [hunk 5]
```diff
@@ -80,19 +127,21 @@ fn main() {
     for i in 0..size {
         xloc[i / VLEN].0[i % VLEN] = i as f64 * inv - 1.5;
     }
-    let xloc = Arc::new(xloc);
-    let pool = CpuPool::new_num_cpus();
+    let xloc2 = xloc.iter().map(|&x| x * x).collect::<Vec<_>>();
```
Comment (TeXitoi) [inline src/mandelbrot.rs]: having a vec of tuple instead of zipping xloc and xloc2 may improve locality, and thus cache usage.

**Comment by cristicbz** [comment]
I just committed random stuff, I think I messed up my git aliases, sorry
:(... I'll fix it when I get home tonight.

File: src/mandelbrot.rs (hunk 5 above)
Comment (cristicbz) [reply src/mandelbrot.rs]: I had tried this; it doesn't help for speed, but I guess it looks nicer.

**Comment by cristicbz** [comment]
Ok, apologies for previous commit, I managed to merge a whole bunch of random experiments stuff somehow. I cleaned it up and pushed, have a look please.

File: src/mandelbrot.rs (hunk 2 above)
Comment (cristicbz) [reply src/mandelbrot.rs:84]: (as mentioned, I tried this but it doesn't help :/)

File: src/mandelbrot.rs [hunk 6]
```diff
@@ -3,15 +3,13 @@
 //
 // contributed by Matt Watson
 // contributed by TeXitoi
+// contributed by Cristi Cobzarenco (@cristicbz)
```
Comment (TeXitoi) [inline src/mandelbrot.rs:6]: feel free to let it like that if you want to submit the bench yourself, or let my name at the end if you want I take care of the submission.

**Review by TeXitoi — APPROVED** [review]

**Comment by cristicbz** [comment]
I can do the submission tomorrow, thanks!

**Comment by TeXitoi** [comment]
OK, please post a link here to the issue when you create it.

**Comment by cristicbz** [comment]
https://alioth.debian.org/tracker/index.php?func=detail&aid=315634&group_id=100815&atid=413122

**Pull request merged**